design/tla/SwarmKit.tla

```diff
@@ -0,0 +1,654 @@
+   changing the configuration or a worker node going down) are tagged with `CountEvent'.
+   Any run of the system cannot have more than `maxEvents' such events.
+
+   See `EventCounter.tla' for details.
+
+`^ \textbf{Preventing certain failures} ^'
+
+   If you're not interested in some actions then you can block them. For example,
+   adding these two constraints in the ``Action Constraint'' box of the
+   ``Advanced Options'' tab tells TLC not to consider workers going down or
+   workers rejecting tasks as possible actions:
+
+   /\ ~WorkerDown
+   /\ ~RejectTask
+*)
+
+(*
+`^ \textbf{Combining task states} ^'
+
+   A finished task can be either in the `complete' or `failed' state, depending on
+   its exit status. If we have 4 finished tasks, that's 16 different states. For
+   modelling, we might not care about exit codes and we can treat this as a single
+   state with another definition override:
+
+   failed <- complete
+
+   In a similar way, we can combine { assigned, accepted, preparing, ready } into a single
+   state:
+
+   accepted <- assigned
+   preparing <- assigned
+   ready <- assigned
+*)
+
+---------------------------- MODULE User  --------------------------------------------
+\* Actions performed by users
+
+(* Create a new service with any ServiceSpec.
+
+   This says that a single atomic step of the system from an old state
+   to a new one is a CreateService step iff `tasks' and `nEvents' don't change
+   and the new value of `services' is the same as before except that some
+   service ID that wasn't used in the old state is now mapped to some
+   ServiceSpec.
+
+   Note: A \ B means { x \in A : x \notin B } --
+         i.e. the set A with all elements in B removed.
+   *)
+CreateService ==
+  /\ UNCHANGED << tasks, nodes, nEvents >>
+  /\ \E sid \in ServiceId \ DOMAIN services,     \* `sid' is an unused ServiceId
+       spec \in ServiceSpec :                    \* `spec' is any ServiceSpec
+          /\ spec.remove = FALSE                 \* (not flagged for removal)
+          /\ services' = services @@ sid :> spec \* Add `sid |-> spec' to `services'
+
+(* Update an existing service's spec. *)
+UpdateService ==
+  /\ UNCHANGED << tasks, nodes >>
+  /\ CountEvent \* Flag as an event for model-checking purposes
+  /\ \E sid     \in DOMAIN services,   \* `sid' is an existing ServiceId
+        newSpec \in ServiceSpec :      \* `newSpec' is any `ServiceSpec'
+       /\ services[sid].remove = FALSE \* We weren't trying to remove sid
+       /\ newSpec.remove = FALSE       \* and we still aren't.
+       \* You can't change a service's mode:
+       /\ (services[sid].replicas = global) <=> (newSpec.replicas = global)
+       /\ services' = [ services EXCEPT ![sid] = newSpec ]
+
+(* The user removes a service.
+
+   Note: Currently, SwarmKit deletes the service from its records immediately.
+   However, this isn't right because we need to wait for the resources to be freed.
```

Comment:
@dperny mentioned that there are also service-level resources that need to be freed. How about:

> However, this isn't right because we need to wait for service-level resources such as Virtual IPs to be freed.